clang/runtime/dpct-rt/include/dpct/device.hpp

```diff
@@ -662,7 +663,14 @@ class dev_mgr {
   void select_device(unsigned int id) {
     std::lock_guard<std::recursive_mutex> lock(m_mutex);
     check_id(id);
-    _thread2dev_map[get_tid()] = id;
+    auto curr_tid = get_tid();
+
+    // make sure the device stack exist
+    auto it = _thread2dev_map.find(curr_tid);
+    if (it == _thread2dev_map.end())
+      _thread2dev_map[curr_tid] = std::stack<unsigned int>();
+
+    _thread2dev_map[curr_tid].push(id);
```

Comment:
There are two unnecessary search operations for the map introduced by the operator[].
How about changing to:
```suggestion
    if (it != _thread2dev_map.end())
      it->second.push(id);
    else {
      std::stack<unsigned int> s;
      s.push(id);
      _thread2dev_map.insert(id);
    }
```

Reply:
Agreed. Code modified.